### ISSUE TITLE
Monitor application log files

### ISSUE THREAD

**Issue by markptrueman** [netdata/netdata#3729]
Parent Of: #4318 

Hi

Is there any way to be able to monitor an application specific log file and then create a chart based on occurences of a log line (like amount of occurences in a day, or a line graph of when the occurences happen). Would like to match the lines on a regex. (This is similar to splunk functionality)

**Comment by l2isbad** [comment]
> application specific log file and then create a chart based on occurences of a log line

Here you go

<details><summary>module example</summary>
<p>

```python
# -*- coding: utf-8 -*-

import re

from bases.FrameworkServices.LogService import LogService

# default module values (can be overridden per job in `config`)
# update_every = 2
priority = 60000
retries = 60

ORDER = ['matched']

CHARTS = {
    'matched': {
        'options': [None, 'Matched Log Line', 'lines', 'matched', 'whatever.matched', 'line'],
        'lines': [
            ['matched', 'matched', 'incremental'],
        ]
    }
}


class Service(LogService):
    def __init__(self, configuration=None, name=None):
        LogService.__init__(self, configuration=configuration, name=name)
        self.order = ORDER
        self.definitions = CHARTS
        self.log_path = self.configuration.get('log_path')
        self.re = self.configuration.get("regex")
        self.data = {"matched": 0}

    def check(self):
        if not LogService.check(self):
            return False

        if not self.re:
            self.error("regex not specified")
            return False

        try:
            self.re = re.compile(self.re)
        except re.error as err:
            self.error("regex compile error: ", err)
            return False

        return True

    def get_data(self):
        raw = self._get_raw_data()

        if not raw:
            return None if raw is None else self.data

        for line in raw:
            self.data['matched'] += bool(self.re.search(line))

        return self.data

```
</p>
</details>

**Comment by Ferroin** [comment]
@l2isbad Any chance you could submit that as a PR?  I would suggest logwatch for a module name, but there's already some software that does this with the same name, so that might get confusing...

**Comment by markptrueman** [comment]
Please excuse my ignorance on this. I've looked at how you set up a plugin and it's a bit confusing. Do i basically create a file in the plugins.d directory (lets call it logwatch.plugin) and then a corresponding conf file in the python.d folder?

**Comment by l2isbad** [comment]
> Any chance you could submit that as a PR?

No chance :smile: In general i like the idea `generic log plugin`. I think we need to think better about how this should be implemented.

**Comment by l2isbad** [comment]
@markptrueman

Do 1 and 2 from https://github.com/firehol/netdata/issues/3729#issuecomment-389864070

`<whatever>.conf example`

```yaml

job_name:
    log_path: "PATH_TO_LOG_FILE"
    regex: "REGULAR_EXPRESSION"
```

**Comment by markptrueman** [comment]
I don't think i'll be submitting any PRs having only been using netdata for 2 days :)

**Comment by Ferroin** [comment]
Personally, I think the generic outline from above is fine.  THe only big improvements I can think of are:

* Support for the user to configure at least the name of the chart and dimensions (and possibly also the context).  It would be very tedious to have to scroll through dozens of charts with the exact same name to find exactly what you want, especially if the dimensions are all the same name too.
* Support for multiple dimensions per chart with different regexes for each so that each monitored log file can have one chart, instead of needing one per regex.

With that, and some basic documentation of Python regexes (or possibly just a link to the online documentation for them) in the config file, it should cover what most people who would want this need.

**Comment by l2isbad** [comment]
> With that, and some basic documentation of Python regexes 

imo [link to official docs](https://docs.python.org/3/library/re.html) is enough

> THe only big improvements I can think of are:

Exactly. I believe list should be longer. @ktsaou please add `discussion` badge.

**Comment by l2isbad** [comment]
Ok ok, 2 links (py2, py3) :smile:

**Comment by Ferroin** [comment]
Yeah, that would work, provided we mention whether we're doing Unicode or ASCII matching by default (because that can have a huge impact on how the matching happens).

Thinking a bit further regarding config, I think having one job and one chart per log file probably makes the most sense, and then allowing some arbitrary number of named dimensions each with its own regex.  Perhaps something like this for config:

```yaml
job_name:
    log_path: /path/to/logfile
    matches:
        dim1_name: <regex1>
        dim2_name: <regex2>
        dim3_name: <regex3>
```

**Comment by markptrueman** [comment]
![logwatch](https://user-images.githubusercontent.com/16737331/40184433-2dd4d4fc-59e8-11e8-9ea7-325eded9a2e9.png)

That looks to be working. I'll see what other features I can add soon :) Thanks guys. I'll drop anything else i think of/implement in here.

**Comment by l2isbad** [comment]
Or even
```yaml
job_name:
    log_path: /path/to/logfile
    matches:
        chart1_name:
            dim1_name: <regex1>
            dim2_name: <regex2>
            dim3_name: <regex3>
        chart2_name:
            dim1_name: <regex1>
            dim2_name: <regex2>
            dim3_name: <regex3>
```

The problem i see is do n regex searches per line. How we can optimize it? :thinking: 

@Ferroin have a look
https://github.com/firehol/netdata/issues/2328#issuecomment-308389267

> '.*.(?:microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)'
> 
> is equal to
> 
> '\.(?:microsoft.com|azure.com|office365.com|office.com|AAA.gr|XXX.com|YYY.com)'
> 
> but 2nd is much faster

**Comment by Ferroin** [comment]
In theory, we could at least optimize for non-matching lines by combining all the regexes into one and checking if a line matches that (it won't get us much, but it would be faster than multiple calls).

Pre-compiling the regexes (with `re.compile()`) would probably be the biggest performance improvement over the example you posted for cases where there's lots of regular expressions.

I've found some assorted stuff on StackOverflow that seems to suggest that there's not really any good option for optimizing multiple matches unless you just need to know what the matched text was or if there were any matches:
https://stackoverflow.com/questions/13030459/match-text-against-multiple-regex-in-python
https://stackoverflow.com/questions/38161309/how-to-match-a-string-against-multiple-regex

If we're going to be picky about performance, re2 may be of interest instead of just using the built-in Python `re` module (see https://github.com/google/re2 and https://github.com/facebook/pyre2/), even though it would add a new dependency.

As far as the efficiency of the regular expressions themselves, I'm not sure we should worry too much about that.  It's not really Netdata's job to make sure the user knows what they're doing (and while the whole `.*.` versus `.` thing does have an impact, it's not likely to be big enough to matter unless they're matching huge numbers of regexes that do that, and most of the other efficiency optimizations that can b e done are similar).

**Comment by l2isbad** [comment]
Actually example i posted has `re.compile()`. I think we have to distinguish somehow if dimension is regex or not. `substring in string` is much faster and ok in many cases.

**Comment by Ferroin** [comment]
@l2isbad Oh, I missed that part of `check()`.  As far as differentiating simple substring matching (and possibly optionally prefix matching with `string.startswith(substring)`, as that is even faster than `string in substring`), that could be as simple as just having some prefix on the regex/string that we strip off (IOW, our own syntax extension).

@markptrueman Unfortunately, that gets a bit more complicated, and is probably better served by writing a plugin specific to whatever application is being monitored.  Note that I'm not saying that it's not possible, but doing it generically would  mean pushing most of the stuff that's typically done in a module into user configuration, which would complicate configuration greatly, and might discourage people from writing new modules.

**Comment by ktsaou** [comment]
About the discussion you had above for the generic log file parser, I would do something like this:

regex supports "Named Groups". We could use regex Named Groups to define variables that can be used for dimensions. We need at least a type.

Something like this:

```yaml
job_name:
    log_path: /path/to/logfile
    regex: just_one_regex_with_named_groups
    chart_name_1:
        dim1_name: named_group_1
        dim2_name: named_group_2
        dim2_name: named_group_3
    chart_name_2:
        dim1_name: named_group_4
        dim2_name: named_group_5
        dim2_name: named_group_6
```

or even:

```yaml
job_name:
    log_path: /path/to/logfile
    regex: just_one_regex_with_named_groups
    chart:
        name: mychart1
        context: mycontext
        family: family
        title: mytitle
        dimensions:
             dimension:
                  name: mydim1
                  type: gauge
                  multiplier: 1
                  divider: 1
             dimension:
                  name: mydim2
                  type: gauge
                  multiplier: 1
                  divider: 1
    chart:
        name: mychart2
        context: mycontext
        family: family
        title: mytitle
        dimensions:
             dimension:
                  name: mydim1
                  type: gauge
                  multiplier: 1
                  divider: 1
             dimension:
                  name: mydim2
                  type: gauge
                  multiplier: 1
                  divider: 1
```

(I am not sure yaml supports the same key multiple times, but I think it supports arrays somehow)

What do you think?

**Comment by l2isbad** [comment]
Looks like log plugins is always more or less specific. I mean user almost always need to add some parse logic.

**Comment by ktsaou** [comment]
> Looks like log plugins is always more or less specific. I mean user almost always need to add some parse logic.

Yes, it is. The whole idea is to check if you can provide a mechanism that is significantly easier to configure than writing specific parsers for each application log.

This what I tried above. For example, `type` could support `set` (to count uniques), and the `value` could accept expressions. I think that the concept of statsd (gauges, counters, histograms and sets) is quite generic for monitoring all kinds of things.

The ideal solution should allow us to ship configs for the default log files of everything and be easily customizable for parsing variations of them.

Generally, you should avoid considering yourself as a reference. For you it will always be easier to write the damn thing directly. But for most people, touching a python file is a no-go...

---

What would be interesting is the possibility of providing an interface for plugins to push statsd metrics via the pipe. netdata statsd already supports synthetic charts (so that metrics can be arranged in charts) and I plan to enhance this a lot (ie. allow expressions, aggregations, etc).

So, log parsing plugins could push just standalone metrics to netdata, netdata would just pass them to its statsd and then synthetic charts could be configured via `/etc/netdata/statsd.d/` to build the actual charts... btw, netdata statsd over pipe is fast, very fast. Something like 4mil metrics per second per core.

This could  be extended further: convert the internal database of netdata to "just a bunch of metrics" and de-couple charts grouping from data-collection. So, all plugins will just push individual metrics to netdata, and synthetic charts could organize them into charts, applications, etc. Seems like a re-write of the whole thing though...

**Comment by Ferroin** [comment]
@ktsaou I actually like that idea a bit better.  The only complicating factor I can see is that while YAML supports multiple keys with the same name, most parsers don't (and PyYAML in particular doesn't, subsequent keys with the same name override previous ones, so only the last key with a given name shows up in the resultant parsed structure).

**Comment by l2isbad** [comment]
> The whole idea is to check if you can provide a mechanism that is significantly easier to configure than writing specific parsers for each application log.

Sounds good, but doesn't look achievable to me (`significantly easier to configure than writing`).

 Even 2nd config example from https://github.com/firehol/netdata/issues/3729#issuecomment-392210804 looks complicated to me (from the user side) without

> type could support set (to count uniques), and the value could accept expressions

This is where `significantly easier to configure than writing` fails.

Sorry for the pessimism.

**Comment by ktsaou** [comment]
> Sorry for the pessimism.

You are probably right. To do it right, a new complex tool has to be created...

**Comment by Ferroin** [comment]
This is probably a case where it just needs to be done well enough, not perfectly.  Even supporting just basic regexes  with a syntax like in the first example from that comment is still vastly simpler for _most_ people than writing up a custom plugin for it, especially if you just want to count matching lines directly and not do some kind of processing.

Unless there's some particular objection, I can look at putting together something simple to do this.

**Comment by l2isbad** [comment]
This is the only request for matching log line. 
The guy mention `This is similar to splunk functionality`, maybe we should look what is it :smile:

**Comment by Ferroin** [comment]
Well yeah, there's a bunch of log analyzers out there.  The big thing here is that pretty much none of them support the near-real-time type monitoring that Netdata does (they're designed for infrequent batch processing), so there would be a significant advantage to having something like this that can fire alerts on certain matched lines.

**Comment by Thijzer** [comment]
I completely agree and am looking for this. We have applications running with lots of log files on different debug levels. Some of these need to be monitored and have triggers. Sorry for the bump.

**Comment by ktsaou** [comment]
Monitoring application log files is a quite generic need.

We have discussed in the past (above) with @l2isbad and @Ferroin the possibility of providing a universal plugin for tailing log files and configuring a set of regular expressions to parse it. We then concluded that the effort of this is too much, compared to adapting a simple python module.

@cakrit until we provide such a universal log parsing plugin, we should at least provide detailed documentation on how to accomplish this with a custom python module.
